zapcore/write_syncer.go

```diff
@@ -75,6 +78,87 @@ func (s *lockedWriteSyncer) Sync() error {
 	return err
 }
 
+type bufferWriterSyncer struct {
+	ws           WriteSyncer
+	bufferWriter *bufio.Writer
+}
+
+// defaultBufferSize sizes the buffer associated with each WriterSync.
+const defaultBufferSize = 256 * 1024
+
+// defaultFlushInterval means the default flush interval
+const defaultFlushInterval = 30 * time.Second
+
+// CancelFunc should be called when the caller exits to clean up buffers.
+type CancelFunc func() error
+
+// Buffer wraps a WriteSyncer in a buffer to improve performance,
+// if bufferSize = 0, we set it to defaultBufferSize
+// if flushInterval = 0, we set it to defaultFlushInterval
+func Buffer(ws WriteSyncer, bufferSize int, flushInterval time.Duration) (WriteSyncer, CancelFunc) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	cancelfunc := func() error {
+		cancel()
```

Comment:
after `cancel`, can we wait for the background goroutine to end?

Reply:
cancel would close the channel, and we do not need to wait for the goroutine to end apparently.

see the go comments below:

// The close built-in function closes a channel, which must be either
// bidirectional or send-only. It should be executed only by the sender,
// never the receiver, and has the effect of shutting down the channel after
// the last sent value is received. **After the last value has been received
// from a closed channel c**, any receive from c will succeed without
// blocking, returning the zero value for the channel element. The form
//	x, ok := <-c
// will also set ok to false for a closed channel.

Reply:
We don't need to wait for it to end, but I think it's a better API for callers that once `CloseFunc` returns, the "close" work is done.

E.g., if someone calls `defer CloseFunc` in their `main` function, we want to guarantee that we've done writing to the background